ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -658,33 +834,55 @@ IndexValue markAsDeleted(StoreKey id, FileSpan fileSpan, long deletionTimeMs) th
+  IndexValue markAsDeleted(StoreKey id, FileSpan fileSpan, MessageInfo info, long deletionTimeMs, short lifeVersion)
       throws StoreException {
+    boolean hasLifeVersion = IndexValue.hasLifeVersion(lifeVersion);
     validateFileSpan(fileSpan, true);
     IndexValue value = findKey(id);
     if (value == null && info == null) {
       throw new StoreException("Id " + id + " not present in index " + dataDir, StoreErrorCodes.ID_Not_Found);
-    } else if (value != null && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
-      throw new StoreException("Id " + id + " already deleted in index " + dataDir, StoreErrorCodes.ID_Deleted);
+    } else if (value != null) {
+      if (hasLifeVersion) {
+        // When this method is invoked in either recovery or replication, delete can follow any index value.
+        if ((value.isDelete() && value.getLifeVersion() >= lifeVersion) || (value.getLifeVersion() > lifeVersion)) {
+          throw new StoreException("LifeVersion conflict in index. Id " + id + " LifeVersion: " + value.getLifeVersion()
+              + " Delete LifeVersion: " + lifeVersion, StoreErrorCodes.Life_Version_Conflict);
+        }
+      } else {
+        if (value.isDelete()) {
+          throw new StoreException("Id " + id + " already deleted in index " + dataDir, StoreErrorCodes.ID_Deleted);
+        }
+      }
     }
     long size = fileSpan.getEndOffset().getOffset() - fileSpan.getStartOffset().getOffset();
     IndexValue newValue;
     if (value == null) {
       // It is possible that the PUT has been cleaned by compaction
+      if (!hasLifeVersion) {
+        throw new StoreException("MessageInfo of delete carry invalid lifeVersion",
+            StoreErrorCodes.Initialization_Error);
+      }
       newValue =
-          new IndexValue(size, fileSpan.getStartOffset(), info.getExpirationTimeInMs(), info.getOperationTimeMs(),
-              info.getAccountId(), info.getContainerId());
+          new IndexValue(size, fileSpan.getStartOffset(), IndexValue.FLAGS_DEFAULT_VALUE, info.getExpirationTimeInMs(),
+              info.getOperationTimeMs(), info.getAccountId(), info.getContainerId(), lifeVersion);
       newValue.clearOriginalMessageOffset();
     } else {
+      lifeVersion = hasLifeVersion ? lifeVersion : value.getLifeVersion();
       newValue =
           new IndexValue(value.getSize(), value.getOffset(), value.getFlags(), value.getExpiresAtMs(), deletionTimeMs,
-              value.getAccountId(), value.getContainerId(), (short) 0);
+              value.getAccountId(), value.getContainerId(), lifeVersion);
       newValue.setNewOffset(fileSpan.getStartOffset());
+      // Only set the original message offset when the value is put
+      if (!value.isPut()) {
+        newValue.clearOriginalMessageOffset();
+      }
       newValue.setNewSize(size);
     }
+    newValue.clearFlag(IndexValue.Flags.Undelete_Index);
```

Comment:
should we clear TTL Update flag as well?

Reply:
we don't, a delete or undelete can carry ttl_update_index flag and it's expiration date, since a delete and a undelete should represent the final state of a key. And expiration date is the final state of this key.